Improved sub-routine for string `quantity` arguments for find_unit

Shikhar1998 opened the pull request:
Re-written sub-routine for string quantity arguments for find_unit.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
Fixes #13874
#### Brief description of what is fixed or changed
The input element is searched with all the elements of `units` dictionary and matched for their`dimensions`. 
#### Other comments
Moreover, in PR #13438 problem of multiple instances of `plank_length` and `plank_charge` in resultant list for  `find_unit('charge')` and `find_unit('length')` was raised. 

Example:
```
>>> u.find_unit('charge')
['C', 'coulomb', 'coulombs', 'planck_charge', 'planck_charge']
```
To counter this problem, `set` was used. However, there is no special need for it, as all the `definitions` are inherently unique. This problem has also been eliminated.

Shikhar1998 committed:
Fixes find_unit

Re-written sub-routine for string quantity arguments for find_unit.

Shikhar1998 committed:
Some changes in tests

Now covers all possible cases

ylemkimon commented on sympy/physics/units/__init__.py:
```diff
@@ -222,23 +222,28 @@ def find_unit(quantity):
 
```
I think docstring should explain its exact behavior

Shikhar1998 replied:
@ylemkimon, I have made the requested changes, please have a look.

ylemkimon replied:
I think `all instances of units/dimensions with these arguments.` is little ambiguous and it doen.
If a string is given, it currently returns
(i) all units matching the string
(ii) if a dimension exactly matching the string exists, all units with that dimension
(iii) if a unit exactly matching the string exists, all units with same dimension

ylemkimon reviewed (changes requested):
+1

Shikhar1998 committed:
Improved Documentation

Explained the result for possible 2 input cases: 1. String and 2. Quantity/Dimension

Shikhar1998 committed:
Formatting Docstring

72 character rule

Shikhar1998 commented:
@smichr, can you please review this PR.

ylemkimon commented on sympy/physics/units/__init__.py:
```diff
@@ -210,35 +210,46 @@
+        if isinstance(dim, Quantity): dim = dim.dimension
+        for i in dir(u):
+            quant_attr = getattr(u, i)
+            if quantity in i or (isinstance(quant_attr, Quantity) and quant_attr.dimension == dim):
```
I don't think it should return `Dimension`s.

Shikhar1998 replied:
True, as name of function `find_unit` itself implies to find all instances of that unit. However, original `docstring` states that function should return all `Dimensions`.

Shikhar1998 replied:
A separate function like `find_dimension` can be added to return all the dimensions. Please suggest.

Shikhar1998 replied:
However, the results of these functions (`find_unit` and `find_dimension`) would be quite alike, which may make this exercise quite dispensable.

Shikhar1998 committed:
Removed definations from output

Shikhar1998 commented:
Done!! ping @ylemkimon .

Shikhar1998 commented:
@ylemkimon, please recommend any other changes as soon as you get time.

Shikhar1998 commented:
@ylemkimon, ping.

ylemkimon commented:
I think documentation needs improvement, but I'm not a native English speaker, so I cannot make a suggestion.

Shikhar1998 commented:
@asmeurer, can you please suggest some changes in the documentation.

Shikhar1998 committed:
Modifications in Documentation

Made changes as suggested by @smichr

Shikhar1998 commented on sympy/physics/units/__init__.py:
```diff
@@ -210,12 +210,18 @@
 
 def find_unit(quantity):
     """
-    Return a list of matching units or dimension names.
```
Quite a nice idea, thanks for the insight. Made changes, please check.

smichr commented on sympy/physics/units/__init__.py:
```diff
@@ -210,35 +210,54 @@
     >>> u.find_unit(u.inch**3)[:5]
     ['l', 'cl', 'dl', 'ml', 'liter']
+
+    # When the input is a string, it returns units which have input
```
Two comments:
* go ahead and write the comments without the leading `#`
* the last comment about using strings is a bit confusing, perhaps
because of 'sub-script'. I think "substring" would be more recognized in this context. Myabe
```
When the input is a string, units or Quantities are returned. The units will
have the input as a substring; Quantities will a dimension that matches the
input.
```

But if that is right, I'm not sure why the behavior of trying to match dimensions was added. That can already be done with u.dimension. string input is used when you aren't sure of the spelling or quantity that you are looking for; once you find it you can then use the corresponding dimension in your next search. Is there an unecessary complication by combining both outputs for a string input? e.g. if you can't do `[u.find_unit(gettattr(u, i)) for i in u.find_unit('foo')]` ... or soemthing like that ... then this is maybe not a good direction to go.

Shikhar1998 replied:
> That can already be done with u.dimension. string input is used when you aren't sure of the spelling or quantity that you are looking for; once you find it you can then use the corresponding dimension in your next search.

Hmm.., I have a simple idea for this. We can match the various quantities in the dictionary `u` for given string and gives output like this:   
```
 >>> u.find_unit('char')
did you mean: charge                          # for incomplete enteries
 >>> u.find_unit('charge')
1                                               # a perfect match
 >>> u.find_unit('0x00x0')
0                                               # no match

smichr commented:
What do you think about these changes:

```diff
diff --git a/sympy/physics/units/__init__.py b/sympy/physics/units/__init__.py
index 75de4ee..dc38476 100644
--- a/sympy/physics/units/__init__.py
+++ b/sympy/physics/units/__init__.py
@@ -249,29 +249,38 @@ def find_unit(quantity):
 
     """
     import sympy.physics.units as u
-    rv = []
+    rv = set()
+    IA = [(i, getattr(u, i)) for i in dir(u)]
+    IA = [(i, a) for i, a in IA if isinstance(a, (Quantity, Dimension))]
     if isinstance(quantity, string_types):
-        dim = getattr(u, quantity)
-        if isinstance(dim, Quantity): dim = dim.dimension
-        for i in dir(u):
-            quant_attr = getattr(u, i)
-            if quantity in i or (isinstance(quant_attr, Quantity) and quant_attr.dimension == dim):
-                if not isinstance(quant_attr, Dimension): rv.append(i)
-
+        dim = getattr(u, quantity, None)
+        if dim is None:
+            for i, a in IA:
+                if quantity in i:
+                    rv.add(i)
+                    rv.update(find_unit(i))
+        else:
+            rv.update([i for i, _ in IA if quantity in i])
+            if isinstance(dim, Quantity):
+                dim = dim.dimension
+            for i, quant_attr in IA:
+                if quantity in i or (isinstance(quant_attr, Quantity)
+                        and quant_attr.dimension == dim):
+                    if not isinstance(quant_attr, Dimension):
+                        rv.add(i)
     else:
-        for i in sorted(dir(u)):
-            other = getattr(u, i)
-            if not isinstance(other, Quantity):
-                continue
+        for i, other in IA:
             if isinstance(quantity, Quantity):
-                if quantity.dimension == other.dimension:
-                    rv.append(str(i))
+                if isinstance(other, Quantity) and quantity.dimension == other.dimension:
+                    rv.add(i)
             elif isinstance(quantity, Dimension):
-                if other.dimension == quantity:
-                    rv.append(str(i))
-            elif other.dimension == Dimension(Quantity.get_dimensional_expr(quantity)):
-                rv.append(str(i))
-    return sorted(rv, key=len)
+                if isinstance(other, Quantity) and other.dimension == quantity:
+                    rv.add(i)
+            elif isinstance(other, Quantity) and other.dimension == Dimension(Quantity.get_dimensional_expr(quantity)):
+                rv.add(i)
+    if not rv:
+        return []
+    return list(list(zip(*sorted(zip([len(i) for i in rv], rv))))[1])
 
 # NOTE: the old units module had additional variables:
 # 'density', 'illuminance', 'resistance'.
diff --git a/sympy/physics/units/tests/test_quantities.py b/sympy/physics/units/tests/test_quantities.py
index fd9f6b0..94bda98 100644
--- a/sympy/physics/units/tests/test_quantities.py
+++ b/sympy/physics/units/tests/test_quantities.py
@@ -221,6 +221,8 @@ def test_find_unit():
     assert find_unit('coulomb') == ['C', 'coulomb', 'coulombs',
                                     'planck_charge', 'coulomb_constant']
     assert find_unit(coulomb) == ['C', 'coulomb', 'coulombs', 'planck_charge']
+    assert find_unit('char') == ['C', 'charge', 'coulomb', 'coulombs', 'planck_charge']
+    assert find_unit('charge') == ['C', 'charge', 'coulomb', 'coulombs', 'planck_charge']
     assert find_unit(charge) == ['C', 'coulomb', 'coulombs', 'planck_charge']
     assert find_unit(inch) == [
         'm', 'au', 'cm', 'dm', 'ft', 'km', 'ly', 'mi', 'mm', 'nm', 'pm', 'um',
@@ -237,7 +239,7 @@ def test_find_unit():
         'l', 'cl', 'dl', 'ml', 'liter', 'quart', 'liters', 'quarts',
         'deciliter', 'centiliter', 'deciliters', 'milliliter',
         'centiliters', 'milliliters', 'planck_volume']
-    assert find_unit('voltage') == ['V', 'v', 'volt', 'volts', 'planck_voltage']
+    assert find_unit('voltage') == ['V', 'v', 'volt', 'volts', 'voltage', 'planck_voltage']
 
 
 def test_Quantity_derivative():
```

Shikhar1998 committed:
Modifications in __init__.py

Changes made as suggested by @smichr

Shikhar1998 commented:
@smichr, Thank's for the valuable input. I have committed these suggested changes. I had one question regarding the current implementation; when `u.find_unit('')` is executed, it returns the entire units dictionary. Is this a valid behavior or not?

ylemkimon commented:
@smichr 
> `IA = [(i, a) for i, a in IA if isinstance(a, (Quantity, Dimension))]`

I still think `Dimension`s should not be included in the return value of `find_unit`.

smichr commented:
> I still think Dimensions should not be included in the return value of find_unit.

Can you give an example of what the output is that you would like it not to be?

smichr commented:
> Is this a valid behavior or not?

probably not...we could just special case that and return [].

Shikhar1998 committed:
Handled case: quantity equal to empty string

Now returns an empty list

smichr commented:
@ylemkimon , why don't you think Dimensions should be included in the return result?

As it is now, the original intent of the docstring should be restored to indicate that dimensions may be returned, too. Otherwise, if there is a good reason not to, then only the code needs to be modified and the docstring left alone.

smichr commented:
Would it make sense to return dimensions if the quantity is ''?

```diff
--- a/sympy/physics/units/__init__.py
+++ b/sympy/physics/units/__init__.py
@@ -249,34 +249,27 @@ def find_unit(quantity):

     """
     import sympy.physics.units as u
+    IQ = [(i, getattr(u, i)) for i in dir(u)]
+    if not quantity:
+        return sorted([(i, q) for i, q in IQ if isinstance(q, Dimension)])
+    IQ = [(i, q) for i, q in IQ if isinstance(q, Quantity)]
     rv = set()
-    IA = [(i, getattr(u, i)) for i in dir(u)]
-    IA = [(i, a) for i, a in IA if isinstance(a, (Quantity, Dimension))]
     if isinstance(quantity, string_types):
+        rv.update([i for i, _ in IQ if quantity in i])
         dim = getattr(u, quantity, None)
-        if dim is None:
-            for i, a in IA:
-                if quantity in i:
-                    rv.add(i)
-                    rv.update(find_unit(i))
-        else:
-            rv.update([i for i, _ in IA if quantity in i])
+        if dim is not None:
             if isinstance(dim, Quantity):
                 dim = dim.dimension
-            for i, quant_attr in IA:
-                if quantity in i or (isinstance(quant_attr, Quantity)
-                        and quant_attr.dimension == dim):
-                    if not isinstance(quant_attr, Dimension):
-                        rv.add(i)
-    else:
-        for i, other in IA:
-            if isinstance(quantity, Quantity):
-                if isinstance(other, Quantity) and quantity.dimension == other.dimension:
+            for i, quant in IQ:
+                if dim == quant.dimension:
                     rv.add(i)
-            elif isinstance(quantity, Dimension):
-                if isinstance(other, Quantity) and other.dimension == quantity:
-                    rv.add(i)
-            elif isinstance(other, Quantity) and other.dimension == Dimension(Quantity.get_dimensional_expr(quantity)):
+    else:
+        if isinstance(quantity, Quantity):
+            dim = quantity.dimension
+        else:
+            dim = Dimension(Quantity.get_dimensional_expr(quantity))
+        for i, quant in IQ:
+            if dim == quant.dimension:
                 rv.add(i)
     if not rv:
         return []
```

ylemkimon commented:
@smichr `Dimension` is a dimension of units, rather than a unit, and different class to `Quantity`. The user would expect to perform `Quantity` methods on units returned by `find_unit`.

smichr commented:
@ylemkimon , here's an example of how someone might use the null find to get the dimensions and then use them for a further search:

```
>>> u.find_unit('')
['acceleration', 'action', 'amount', 'amount_of_substance', 'capacitance', 'char
ge', 'conductance', 'current', 'energy', 'force', 'frequency', 'impedance', 'ind
uctance', 'length', 'luminosity', 'luminous_intensity', 'magnetic_density', 'mag
netic_flux', 'magnetic_flux_density', 'mass', 'momentum', 'power', 'pressure', '
speed', 'temperature', 'time', 'velocity', 'voltage', 'volume']

>>> u.find_unit(u.length/u.time)
['sr', 'deg', 'mil', 'rad', 'degree', 'radian', 'degrees', 'percent', 'radians',
 'percents', 'permille', 'steradian', 'steradians', 'angular_mil', 'angular_mils
', 'avogadro_number']
```

...not sure why `avogadro_number` is showing up, though.

Shikhar1998 commented:
@ylemkimon, will it be fine to rename the function as find_unit_dimension(). I think this will the eliminate the possibility of any unnecessary confusion in future.

smichr commented:
> rename the function as find_unit_dimension

Or a function named `dimensions` that simply returns the list sorted list of dimensions.

czgdp1807 commented:
@Upabjojr Is this PR having a chance to be merged?

czgdp1807 commented:
The PR seems to be a valid one. @Shikhar1998 Would you still like to continue your work on this? Thanks for your contributions.
